subsys/usb/device/class/cdc_acm.c

```diff
@@ -223,12 +223,27 @@ static void cdc_acm_write_cb(uint8_t ep, int size, void *priv)
 		k_work_submit_to_queue(&USB_WORK_Q, &dev_data->cb_work);
 	}
 
-	if (ring_buf_is_empty(dev_data->tx_ringbuf)) {
+	/* If size is 0, we want to schedule tx work even if ringbuf is empty to
+	 * ensure that actual payload will not be sent before initialization
+	 * timeout passes.
+	 */
+	if (ring_buf_is_empty(dev_data->tx_ringbuf) && size) {
 		LOG_DBG("tx_ringbuf is empty");
 		return;
 	}
 
-	k_work_schedule_for_queue(&USB_WORK_Q, &dev_data->tx_work, K_NO_WAIT);
+	/* If size is 0, it means that host started polling IN data because it
+	 * has read the ZLP we armed when interface was configured. This ZLP is
+	 * probably the best indication that host has started to read the data.
+	 * Wait initialization timeout before sending actual payload to make it
+	 * possible for application to disable ECHO. The echo is long known
+	 * problem related to the fact that POSIX defaults to ECHO ON and thus
+	 * every application that opens tty device (on Linux) will have ECHO
+	 * enabled in the short window between open() and ioctl() that disables
+	 * the echo (if application wishes to disable the echo).
```

Comment:
It only works once, right?
This is all just in case CDC ACM is used for protocol-less streaming of something like logging (everyone must understand it). And for a sample, not static like hello_world, but logging something periodically, if the user closes and reopens terminal, device all the time connected to the host, the user could still see garbage in terminal. This is probably okay, but IMHO it needs to be documented so we do not have to triage it again.

Also, people have started to think that it is fine not to wait for DTR because this PR will magically fix it, which is obviously not the case. NACKing IN tokens for a period of time and not adding anything to the UART FIFO until DTR are two different things. (e.g. PR #64039, worse, this board now enables a subsystem using an other SYS_INIT, which is not just silly and not necessary but also against the board porting guide). Therefore the commit message should clearly state that it does not release the user from waiting for DTR.

Reply:
> the user closes and reopens terminal, device all the time connected to the host, the user could still see garbage in terminal.

No, that won't happen.  I explained this in one of the earlier PRs that has been linked.  Because the termios settings persist across opens, the race of ECHO being on is only present for the first open.  After that it never be turned on again.

But if the device is removed and then re-connected from USB, without resetting it, then on the reconnect the ZLP delay workaround will not re-trigger the problem will be back.

Reply:
> It only works once, right?

Once per configuration which matches tty device lifetime, which includes termios settings.

> But if the device is removed and then re-connected from USB, without resetting it, then on the reconnect the ZLP delay workaround will not re-trigger the problem will be back.

ZLP does retrigger every time device is configured after it the bus has been reset (which always happens on USB device connection) and therefore the problem won't be back.

Reply:
> No, that won't happen. I explained this in one of the earlier PRs that has been linked. Because the termios settings persist across opens, the race of ECHO being on is only present for the first open. After that it never be turned on again.

Not talking about echo here but I know for sure of one serial port library where that is completely false because I was getting garbage data whenever the port was opened (the port would be at 9600 baud when connected, the program would open it, configure it, set it to 115200 baud but the garbage data from 9600 baud would be emitted. After closing the port, it would revert the settings back to how they were before it opened i.e. 9600 baud, then opening it would give you garbage again), so this is entirely down to applications, just because e.g. picocom does it one way does not mean every tool does it that way

Reply:
> After closing the port, it would revert the settings back to how they were before it opened i.e. 9600 baud, then opening it would give you garbage again), so this is entirely down to applications, just because e.g. picocom does it one way does not mean every tool does it that way

The case where the echo was enabled by any of the applications that did open the port is what we cannot detect in any way. But I believe this is a case that should entirely be treated as host user-space issue, because someone enabled echo on port that shouldn't really have it enabled.

Similarly, there is case where the first application opening tty device does not disable the echo. But in this case this is likely that host user-space really expects the ehco to be on.

Reply:
> But I believe this is a case that should entirely be treated as host user-space issue

Agreed, can't accommodate every possible way that every serial library uses, accommodating to the sensible/widest used ones is good enough.

Reply:
> Not talking about echo here but I know for sure of one serial port library where that is completely false because I was getting garbage data whenever the port was opened

But a serial port library isn't part of POSIX.  termios persisting across open() is part of the POSIX spec, and thus how the Linux kernel works.  And so this race (ECHO here, but baud rate also races on a RS-232 port) can't be fixed by just writing userspace software that works.

A userspace lib setting the port to the wrong baud rate before changing it to the correct one is clearly a library bug, easily fixable.  And if it's saving/restoring, then the baud can be set with `stty` before it starts so it restores the correct value.

I never meant to say in my earlier comment that writing userspace software that does something broken with termios settings was impossible.

The ZLP here isn't really a true fix, it mitigates the race by adding a delay.  There's an unknown delay from when the port is opened to when ioctl() is called to fix the termios settings.  There's a delay from the ZLP being sent, at open(), and Zephyr CDC ACM sending data (configurable value added in this PR).  I think I recently read someone saying that you can't fix a race by juggling delays.  The real fix is a change in sequencing, but it's on the POSIX/Linux side that it needs to be done.  But even though it's just a mitigation of the race and not a fix, it's clearly a huge improvement worth doing IMHO.

Reply:
> > the user closes and reopens terminal, device all the time connected to the host, the user could still see garbage in terminal.
> 
> No, that won't happen. I explained this in one of the earlier PRs that has been linked. Because the termios settings persist across opens, the race of ECHO being on is only present for the first open. After that it never be turned on again.

It could happen, because application may restore tty settings or reset to default by something else.

Reply:
> It could happen, because application may restore tty settings or reset to default by something else.

Configure the port disable echo via automatically when it's first detected via a udev rule.  Then the application doesn't have to do it and an application that restores settings will not change them back to incorrect values.

The application could also be full of bugs and crash randomly too.

Reply:
> > It could happen, because application may restore tty settings or reset to default by something else.
> 
> Configure the port disable echo via automatically when it's first detected via a udev rule. Then the application doesn't have to do it and an application that restores settings will not change them back to incorrect values.
> 
> The application could also be full of bugs and crash randomly too.

Yes, discussed internally today, and then user does not need this patch.

Reply:
> Yes, discussed internally today, and then user does not need this patch.

The user does need this patch, because no matter how you set up the udev rule there'll be time window between device being `open()` and the `ioctl()` that disables the echo. This time window is what the ZLP tries to mitigate.

Reply:
Yes, exactly.  udev is userspace and can not solve the POSIX race of opening the port and starting the ECHO process before making the ioctl() call on the open port to turn it off.

If a userspace app has silly behavior, this can always be fixed in userspace.  One could go as far as LD_PRELOADing a wrapper that modifies the C lib `tcsetattr()` function to inject alternate termios settings.